app/src/main/java/chat/rocket/android/push/gcm/PushInstanceIDListenerService.java

```diff
@@ -0,0 +1,62 @@
+package chat.rocket.android.push.gcm;
+
+import android.content.Intent;
+
+import com.google.android.gms.iid.InstanceIDListenerService;
+
+import java.util.List;
+import chat.rocket.android.model.ServerConfig;
+import chat.rocket.android.model.ddp.PublicSetting;
+import chat.rocket.android.model.ddp.PublicSettingsConstants;
+import chat.rocket.android.realm_helper.RealmHelper;
+import chat.rocket.android.realm_helper.RealmStore;
+
+public class PushInstanceIDListenerService extends InstanceIDListenerService {
+
+  @Override
+  public void onTokenRefresh() {
+    List<ServerConfig> serverConfigs = getServerConfigs();
+
+    flagForRefresh(serverConfigs);
+
+    if (!shouldRefreshToken(serverConfigs)) {
+      return;
+    }
+
+    Intent intent = new Intent(this, RegistrationIntentService.class);
+    startService(intent);
+  }
+
+  private List<ServerConfig> getServerConfigs() {
+    return RealmStore.getDefault().executeTransactionForReadResults(
+        realm -> realm.where(ServerConfig.class).findAll());
+  }
+
+  private void flagForRefresh(List<ServerConfig> serverConfigs) {
```

Comment:
The responsibility of this method cannot be guessed by the name `flagForRefresh`

`updateSyncPushToken(List<ServerConfig>)` or similar one would be better.